[OpenMP] Clang Codegen Interop : Accept multiple init

SunilKuravinakop opened the pull request:
Modifying clang/lib/CodeGen/CGStmtOpenMP.cpp to accept multiple `init` clauses with `interop` directive.

Commit:
Accept multiple init clauses in interop directive.

  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

Commit:
Merge branch 'main' into interop_multiple_init

Commit:
Merge branch 'main' into interop_multiple_init

Commit:
Merge branch 'main' into interop_multiple_init

Commit:
Adding a testcase for interop_codegen.cpp to check "init" clause being
provided twice.
This check is being done only for AMD GPU.

  Changes to be committed:
 	new file:   clang/test/OpenMP/interop_codegen.cpp

Commit:
Added -fopenmp-targets=nvptx64-nvidia-cuda &
-fopenmp-targets=powerpc64le-ibm-linux-gnu.
  Changes to be committed:
 	modified:   clang/test/OpenMP/interop_codegen.cpp

Commit:
Merge branch 'main' into interop_multiple_init

Commit:
Merge branch 'main' into interop_multiple_init

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -7023,19 +7023,25 @@ void CodeGenFunction::EmitOMPInteropDirective(const OMPInteropDirective &S) {
+  auto It = S.getClausesOfKind<OMPInitClause>();
+  if (!It.empty()) {
+    // Look at the multiple init clauses
+    for (auto C : It) {
```
Expand 'auto' here

SunilKuravinakop replied:
For the Iterator variable `It` `auto It` is preferable rather than 
`llvm::iterator_range<clang::OMPExecutableDirective::specific_clause_iterator<clang::OMPInitClause> > It`

alexey-bataev replied:
C is just OMPClause * here, not an iterator

alexey-bataev replied:
I asking about auto for C, not for It.

Commit:
Changing from "for (auto C : It)" to "for (const class OMPInitClause *C : It)".
  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

SunilKuravinakop commented:
Alexey, 
Can you please review my changes?

Commit:
Generation of llvm-IR for multiple init, use and destroy clauses.

  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp

Commit:
Adding code for multiple occurences of "use" and "destroy" clause.
  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp
 	modified:   clang/test/OpenMP/interop_codegen.cpp

alexey-bataev commented on clang/lib/CodeGen/CGStmtOpenMP.cpp:
```diff
@@ -7023,31 +7023,47 @@ void CodeGenFunction::EmitOMPInteropDirective(const OMPInteropDirective &S) {
+                                      Data.HasNowaitClause);
+    }
+  }
+  auto ItOMPDestroyClause = S.getClausesOfKind<OMPDestroyClause>();
+  if (!ItOMPDestroyClause.empty()) {
+    // Look at the multiple destroy clauses
+    for (const OMPDestroyClause *C : ItOMPDestroyClause) {
+      llvm::Value *InteropvarPtr =
+          EmitLValue(C->getInteropVar()).getPointer(*this);
+      OMPBuilder.createOMPInteropDestroy(Builder, InteropvarPtr, Device,
+                                         NumDependences, DependenceList,
+                                         Data.HasNowaitClause);
+    }
+  }
+  auto ItOMPUseClause = S.getClausesOfKind<OMPUseClause>();
+  if (!ItOMPUseClause.empty()) {
+    // Look at the multiple use clauses
+    for (const OMPUseClause *C : ItOMPUseClause) {
+      llvm::Value *InteropvarPtr =
+          EmitLValue(C->getInteropVar()).getPointer(*this);
+      OMPBuilder.createOMPInteropUse(Builder, InteropvarPtr, Device,
+                                     NumDependences, DependenceList,
+                                     Data.HasNowaitClause);
     }
```
Split it to a separate patch

Commit:
Moving the multiple occurences of use and destroy clauses to a separate
merge.
  Changes to be committed:
 	modified:   clang/lib/CodeGen/CGStmtOpenMP.cpp
 	modified:   clang/test/OpenMP/interop_codegen.cpp

alexey-bataev reviewed (approved):
LG
